### PR TITLE
Wip/6.0 element collection

### PR DESCRIPTION
**Pull request by dreab8**

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/sql/results/internal/ResultsHelper.java [hunk 1]
```diff
@@ -117,6 +119,17 @@ public static void finalizeCollectionLoading(
...
+		if ( collectionInstance instanceof PersistentArrayHolder ) {
+			// in case of PersistentArrayHolder we have to realign the EntityEntry loaded state with
+			// the entity values
+			final Object owner = collectionInstance.getOwner();
+			final EntityEntry entry = persistenceContext.getEntry( owner );
+			final int propertyIndex = entry.getPersister()
+					.getEntityMetamodel()
+					.getPropertyIndex( collectionDescriptor.getAttributeMapping().getAttributeName() );
+			final Object[] loadedState = entry.getLoadedState();
+			loadedState[propertyIndex] = entry.getPersister().getValue( owner, propertyIndex );
```
[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/sql/results/internal/ResultsHelper.java:131]: i think the issue with Arrays is that we cannot set the PersistentArrayHolder  as the property value;

[the remark this answers is not in the document]
Comment (beikov) [reply hibernate-core/src/main/java/org/hibernate/sql/results/internal/ResultsHelper.java:131]: I get that, but when we initialize the loaded state array in the entity entry, shouldn't this be taken care of?

[the remark this answers is not in the document]
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/sql/results/internal/ResultsHelper.java:131]: it is what i'm doing now when have the state of the loaded array

File: hibernate-core/src/main/java/org/hibernate/sql/results/internal/ResultsHelper.java [hunk 2]
```diff
@@ -145,6 +147,15 @@ public static void finalizeCollectionLoading(
...
 		if ( collectionDescriptor.getCollectionType().hasHolder() ) {
+			// in case of PersistentArrayHolder we have to realign the EntityEntry loaded state with
+			// the entity values
+			final Object owner = collectionInstance.getOwner();
+			final EntityEntry entry = persistenceContext.getEntry( owner );
+			final int propertyIndex = entry.getPersister()
+					.getEntityMetamodel()
+					.getPropertyIndex( collectionDescriptor.getAttributeMapping().getAttributeName() );
+			final Object[] loadedState = entry.getLoadedState();
+			loadedState[propertyIndex] = entry.getPersister().getValue( owner, propertyIndex );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/results/internal/ResultsHelper.java:158]:
```suggestion
			final StateArrayContributerMapping mapping = (StateArrayContributerMapping) collectionDescriptor.getAttributeMapping();
			final int propertyIndex = mapping.getStateArrayPosition();
			final Object[] loadedState = entry.getLoadedState();
			loadedState[propertyIndex] = mapping.getValue( owner );
```

**Review by beikov — APPROVED** [review]

**Pull request merged**